### PR TITLE
Feature/362 Tags

### PR DESCRIPTION
**Pull request by JVerbruggen**
Tags

## Description

The addition of tags can give users a better insight into certain characteristics of projects. This pull request introduces tags to dex.

## Type of change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [x] This change requires a documentation update

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I did not update API Controllers, if I did, I added/changed Postman tests
-   [x] I have performed a self-review of my own code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have made corresponding changes to the documentation
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes
-   [x] I updated the changelog with an end-user readable description
-   [x] I assigned this pull request to the correct project board to update the sprint board

## Steps to Test or Reproduce

Outline the steps to test or reproduce the PR here.
These steps will be used during release testing.

1. Do something here...
2. Then run this test...
3. After that, you should see the result here

## Link to issue

Closes: #362

### CODE REVIEW

File: API/Controllers/TagController.cs [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        public async Task<IActionResult> GetTag(int tagId)
+        {
+            if(tagId < 0)
+            {
+                ProblemDetails problem = new ProblemDetails
+                                         {
+                                             Title = "Failed getting tag.",
+                                             Detail =
+                                                 "The Id is smaller then 0 and therefore it could never be a valid tag id.",
+                                             Instance = "5024ADDA-6DE2-4B49-896A-526E8EC4313D"
```
Comment (Ruby77) [inline API/Controllers/TagController.cs:83]: Duplicate instance guid.

File: API/Controllers/TagController.cs [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+            Tag tag = await tagService.FindAsync(tagId)
+                                         .ConfigureAwait(false);
+            if(tag == null)
+            {
+                ProblemDetails problem = new ProblemDetails
+                                         {
+                                             Title = "Failed getting tag.",
+                                             Detail = "The tag could not be found in the database.",
+                                             Instance = "1739EFA6-3F31-4C88-B596-74DA403AC51B"
```
Comment (Ruby77) [inline API/Controllers/TagController.cs:96]: Duplicate instance guid. Make sure all the guids are unique.

**Review by Ruby77 — CHANGES_REQUESTED** [review]
Awesome job, congratz on your first PR this semester :tada:! Well done, I can really see you understand the underlying architecture of the application.

Your pull request did not have a description and steps to reproduce. I understand the general idea of this pull request, but it would be nice if you could update the pull request anyways (it's easier for others to review).

Your code looked really good. I found some small notes and made some suggestions, check if you agree with me (if you don't agree, let me know why :wink:). Keep up the good work Jurjen!

File: API/Controllers/TagController.cs [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+                ProblemDetails problem = new ProblemDetails
+                                         {
+                                             Title = "Failed to save the new tag.",
+                                             Detail = "There was a problem while saving the tag to the database.",
+                                             Instance = "D56DBE55-57A1-4655-99C5-4F4ECEEE3BE4"
+                                         };
+                return BadRequest(problem);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (JVerbruggen) [reply API/Controllers/TagController.cs:148]: Fixed by adding RequiredAttribute to TagResource

File: API/Controllers/TagController.cs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    /// </summary>
+    [Route("api/[controller]")]
+    [ApiController]
+    public class TagController : ControllerBase
+    {
+
+        private readonly IMapper mapper;
+        private readonly ITagService tagService;
+        private readonly IProjectTagService projectTagService;
+        private readonly IUserService userService;
```
[the remark this answers is not in the document]
Comment (JVerbruggen) [reply API/Controllers/TagController.cs:29]: Fixed in commit

File: IdentityServer/6_IdentityServer.csproj [hunk 5]
```diff
@@ -6,6 +6,10 @@
         <Version>1.0.1-beta</Version>
     </PropertyGroup>
 
+    <ItemGroup>
+      <Compile Include="..\Data\Helpers\Seed.cs" Link="Seed.cs" />
+    </ItemGroup>
+
```
[the remark this answers is not in the document]
Comment (JVerbruggen) [reply IdentityServer/6_IdentityServer.csproj:12]: This has been generated, I don't know what it does.

File: Repositories/TagRepository.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    ///     This is the interface of the tag repository
+    /// </summary>
+    /// <seealso cref="IRepository{Tag}"/>
+    public interface ITagRepository : IRepository<Tag> {
+
+        /// <summary>
+        ///     This interface method gets all tags asynchronously
+        /// </summary>
+        /// <returns></returns>
+        Task<List<Tag>> GetAllAsync();
```
[the remark this answers is not in the document]
Comment (JVerbruggen) [reply Repositories/TagRepository.cs:37]: Haven't given much thought to it to be honest. Since it uses ToListAsync it is also kind of a fake async.

The reason I did it is that I took inspiration from other repositories, which did the same thing. (RoleRepository)

Would you suggest removing it and using the one from the base class?

**Comment by JVerbruggen** [comment]
@Ruby77 
> Your pull request did not have a description and steps to reproduce. I understand the general idea of this pull request, but it would be nice if you could update the pull request anyways (it's easier for others to review).

What should be considered as steps to reproduce in this case? Reproduce what? Expected functionality?

**Comment by Ruby77** [comment]
> @Ruby77
> 
> > Your pull request did not have a description and steps to reproduce. I understand the general idea of this pull request, but it would be nice if you could update the pull request anyways (it's easier for others to review).
> 
> What should be considered as steps to reproduce in this case? Reproduce what? Expected functionality?

Good question @JVerbruggen! In Steps to Test or Reproduce you explain how to test your code. So in your case, this could be something like this: "Execute the CRUD functionality from the tag controller as admin and validate that other roles are not able to do this. Validate you are able to add, update and delete tags on your own project. Validate that the data officer is able to update/add/delete tags on projects from users from the same institute. And finally, validate that the admin is able to update/add/delete tags on projects from everyone.".

I hope this makes sense, let me know otherwise :wink:

File: Repositories/TagRepository.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ruby77) [reply Repositories/TagRepository.cs:37]: Yes, I think that would be the best solution to keep everything DRY. Maybe the base method has been added later, but I can't say anything for sure about that.

File: IdentityServer/6_IdentityServer.csproj (hunk 5 above)
[the remark this answers is not in the document]
Comment (Ruby77) [reply IdentityServer/6_IdentityServer.csproj:12]: Check if it's removable. Something has triggered to add this to the .csproj file from the identity server. However, I don't see this in use right now, so you could try removing it. If it breaks, well... we know it's in use then :wink:

**Review by niraymak — CHANGES_REQUESTED** [review]
Nicely done! I have a few requests. Good job that you also updated Postman tests and unit tests. 🥇 

We will meet about the system tomorrow to discuss if the implementation matches how we envision it.

**Comment by JVerbruggen** [comment]
- These tags are going to be renamed to categories

File: CHANGELOG.md [hunk 7]
```diff
@@ -18,6 +18,8 @@ Graduation system for users whose FHICT account becomes unavailable. [#287](http
 - Added wizard to automatically import project with a public and an oauth flow - [326](https://github.com/DigitalExcellence/dex-backend/issues/326)
 
 >>>>>>> feature/wizzard
```
[the remark this answers is not in the document]
Comment (JVerbruggen) [reply CHANGELOG.md:20]: Ill fix this by merging develop into here

**Comment by JVerbruggen** [comment]
Ready for next review! 😄

**Comment by JVerbruggen** [comment]
Current postman test error to fix:

```
1.  AssertionError  Category name is set correctly and matches: Some Postman Category                
2021-03-04T17:31:23.7894119Z      iteration: 2    expected 'Some Updated Postman Category' to deeply equal 'Some Postman Category' 
2021-03-04T17:31:23.7929671Z                      at assertion:2 in test-script                                                    
2021-03-04T17:31:23.7931009Z                      inside "ACL / Registered / Category / Category-GetCategory-Registered"           
2021-03-04T17:31:23.7932436Z                                                                                                       
2021-03-04T17:31:23.7933143Z  2.  AssertionError  Tag name is set correctly and matches: Some Postman Category                     
2021-03-04T17:31:23.7934725Z      iteration: 2    expected 'Some Updated Postman Category' to deeply equal 'Some Postman Category' 
2021-03-04T17:31:23.7936341Z                      at assertion:2 in test-script                                                    
2021-03-04T17:31:23.7937957Z                      inside "ACL / Data Officer / Category / Category-GetCategory-DataOfficer"        
2021-03-04T17:31:23.7938785Z                                                                                                       
2021-03-04T17:31:23.7939824Z  3.  AssertionError  Category name is set correctly and matches: Some Postman Category                
2021-03-04T17:31:23.7941470Z      iteration: 2    expected 'Some Updated Postman Category' to deeply equal 'Some Postman Category' 
2021-03-04T17:31:23.7942477Z                      at assertion:2 in test-script                                                    
2021-03-04T17:31:23.7943432Z                      inside "ACL / PR / Category / Category-GetCategory-PR"                           
2021-03-04T17:31:23.7944347Z                                                                                                       
2021-03-04T17:31:23.7945266Z  4.  AssertionError  Category name is set correctly and matches: Some Postman Category                
2021-03-04T17:31:23.7946473Z      iteration: 2    expected 'Some Updated Postman Category' to deeply equal 'Some Postman Category' 
2021-03-04T17:31:23.7947661Z                      at assertion:2 in test-script                                                    
2021-03-04T17:31:23.7948643Z                      inside "ACL / Alumni / Category / Category-GetCategory-Alumni"                   
2021-03-04T17:31:23.7949413Z                                                                                                       
2021-03-04T17:31:23.7950099Z  5.  AssertionError  Category name is set correctly and matches: Some Postman Category                
2021-03-04T17:31:23.7951331Z      iteration: 3    expected 'Some Updated Postman Category' to deeply equal 'Some Postman Category' 
2021-03-04T17:31:23.7952320Z                      at assertion:2 in test-script                                                    
2021-03-04T17:31:23.7953405Z                      inside "ACL / Registered / Category / Category-GetCategory-Registered"           
2021-03-04T17:31:23.7954435Z                                                                                                       
2021-03-04T17:31:23.7955128Z  6.  AssertionError  Tag name is set correctly and matches: Some Postman Category                     
2021-03-04T17:31:23.7956408Z      iteration: 3    expected 'Some Updated Postman Category' to deeply equal 'Some Postman Category' 
2021-03-04T17:31:23.7957564Z                      at assertion:2 in test-script                                                    
2021-03-04T17:31:23.7958747Z                      inside "ACL / Data Officer / Category / Category-GetCategory-DataOfficer"        
2021-03-04T17:31:23.7959570Z                                                                                                       
2021-03-04T17:31:23.7960286Z  7.  AssertionError  Category name is set correctly and matches: Some Postman Category                
2021-03-04T17:31:23.7961410Z      iteration: 3    expected 'Some Updated Postman Category' to deeply equal 'Some Postman Category' 
2021-03-04T17:31:23.7962638Z                      at assertion:2 in test-script                                                    
2021-03-04T17:31:23.7963586Z                      inside "ACL / PR / Category / Category-GetCategory-PR"                           
2021-03-04T17:31:23.7964542Z                                                                                                       
2021-03-04T17:31:23.7965418Z  8.  AssertionError  Category name is set correctly and matches: Some Postman Category                
2021-03-04T17:31:23.7966543Z      iteration: 3    expected 'Some Updated Postman Category' to deeply equal 'Some Postman Category' 
2021-03-04T17:31:23.7967875Z                      at assertion:2 in test-script                                                    
2021-03-04T17:31:23.7969244Z                      inside "ACL / Alumni / Category / Category-GetCategory-Alumni"         
```

File: JobScheduler/ApiRequestHandler.cs [hunk 8]
```diff
@@ -65,7 +65,7 @@ public ApiRequestHandler(IHttpClientFactory factory, Config config)
         public async Task<List<UserTask>> GetExpectedGraduationUsersAsync()
         {
             HttpResponseMessage response =
-                await client.GetAsync("api/UserTask/CreateUserTasks/" + config.JobSchedulerConfig.TimeRange);
+                await client.GetAsync("localhost/UserTask/CreateUserTasks/" + config.JobSchedulerConfig.TimeRange);
             return JsonConvert.DeserializeObject<List<UserTask>>(await response.Content.ReadAsStringAsync());
         }
```
Comment (TimSnoek123) [inline JobScheduler/ApiRequestHandler.cs:71]: Weet niet waar dit voor is maar moet dit ook in de productie env naar localhost staan?

File: Repositories/ProjectRepository.cs [hunk 9]
```diff
@@ -174,10 +178,22 @@ public override async Task<Project> FindAsync(int id)
             //Execute the IQueryable to get a collection of results
             List<Project> projectResults = await queryableProjects.ToListAsync();
 
-            //Redact the user after fetching the collection from the project (no separate query needs to be executed)
-            projectResults.ForEach( project => project.User = RedactUser( project.User ) );
-
```
Comment (TimSnoek123) [inline Repositories/ProjectRepository.cs:179]: Dit was door ramon gedaan in https://github.com/DigitalExcellence/dex-backend/issues/376. Dit moet vlm niet gedelete worden. Misschien zijn de changes een beetje in de war geraakt omdat Ramon vlm de pull request na jou had gemaakt.

**Review by TimSnoek123 — APPROVED** [review]
Ramon die had een pull request voor jou gemaakt die eerder gemerged werd, je moet daarom nog eerst nog een keer develop pullen en pushen.

File: API/Controllers/CategoryController.cs [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="CategoryController" /> class
+        /// </summary>
+        /// <param name="categoryService">The category service which is used to communicate with the logic layer.</param>
+        /// <param name="projectTagService">The project tag service which is used to communicate with the logic layer.</param>
+        /// <param name="mapper">The mapper which is used to convert the resources to the model to the resource result.</param>
+        public CategoryController(ICategoryService categoryService, IProjectCategoryService projectTagService, IMapper mapper)
+        {
+            this.categoryService = categoryService;
+            this.projectCategoryService = projectTagService;
```
Comment (niraymak) [inline API/Controllers/CategoryController.cs:39]: projectCategoryService

File: API/Controllers/CategoryController.cs [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+        private readonly ICategoryService categoryService;
+        private readonly IProjectCategoryService projectCategoryService;
+
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="CategoryController" /> class
+        /// </summary>
+        /// <param name="categoryService">The category service which is used to communicate with the logic layer.</param>
+        /// <param name="projectTagService">The project tag service which is used to communicate with the logic layer.</param>
+        /// <param name="mapper">The mapper which is used to convert the resources to the model to the resource result.</param>
+        public CategoryController(ICategoryService categoryService, IProjectCategoryService projectTagService, IMapper mapper)
```
Comment (niraymak) [inline API/Controllers/CategoryController.cs:36]: projectCategoryService

File: API/Controllers/CategoryController.cs [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+
+        private readonly IMapper mapper;
+        private readonly ICategoryService categoryService;
+        private readonly IProjectCategoryService projectCategoryService;
+
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="CategoryController" /> class
+        /// </summary>
+        /// <param name="categoryService">The category service which is used to communicate with the logic layer.</param>
+        /// <param name="projectTagService">The project tag service which is used to communicate with the logic layer.</param>
```
Comment (niraymak) [inline API/Controllers/CategoryController.cs:34]: projectCategoryService

File: API/Resources/ProjectResourceResult.cs [hunk 13]
```diff
@@ -97,6 +97,11 @@ public class ProjectResourceResult
         /// </summary>
         public bool InstitutePrivate { get; set; }
 
+        /// <summary>
+        ///     This gets or sets the tags belonging to a project.
```
Comment (niraymak) [inline API/Resources/ProjectResourceResult.cs:101]: rename tags to category

File: API/Resources/ProjectResultResource.cs [hunk 14]
```diff
@@ -77,6 +77,11 @@ public class ProjectResultResource
         /// </summary>
         public List<ProjectLikesResourceResult> Likes { get; set; }
 
+        /// <summary>
+        ///     This gets or sets the tags of the project.
```
Comment (niraymak) [inline API/Resources/ProjectResultResource.cs:81]: rename to categories

File: JobScheduler/ApiRequestHandler.cs [hunk 15]
```diff
@@ -65,7 +65,7 @@ public ApiRequestHandler(IHttpClientFactory factory, Config config)
         public async Task<List<UserTask>> GetExpectedGraduationUsersAsync()
         {
             HttpResponseMessage response =
-                await client.GetAsync("api/UserTask/CreateUserTasks/" + config.JobSchedulerConfig.TimeRange);
+                await client.GetAsync("localhost/UserTask/CreateUserTasks/" + config.JobSchedulerConfig.TimeRange);
```
Comment (niraymak) [inline JobScheduler/ApiRequestHandler.cs:68]: Why did you change this? I guess it should be api/UserTask/CreateUserTasks/ since it is communicating with the api.

The base url is set in the program of this application.

File: JobScheduler/ApiRequestHandler.cs [hunk 16]
```diff
@@ -75,7 +75,7 @@ public async Task<List<UserTask>> GetExpectedGraduationUsersAsync()
         /// <param name="userTaskId"></param>
         public void SetGraduationTaskStatusToMailed(int userTaskId)
         {
-            client.PutAsync("api/UserTask/SetToMailed/" + userTaskId, null);
+            client.PutAsync("localhost/UserTask/SetToMailed/" + userTaskId, null);
```
Comment (niraymak) [inline JobScheduler/ApiRequestHandler.cs:78]: Why did you change this? I guess it should be api/UserTask/CreateUserTasks/ since it is communicating with the api.

The base url is set in the program of this application.

File: JobScheduler/ApiRequestHandler.cs (hunk 15 above)
Comment (JVerbruggen) [reply JobScheduler/ApiRequestHandler.cs:68]: I do not remember making this change at all. I will change it back.

File: Data/Migrations/ApplicationDbContextModelSnapshot.cs [hunk 17]
```diff
@@ -1,645 +1,647 @@
...
 
-                                    b.Property<string>("IdentityId")
-                                     .IsRequired()
-                                     .HasColumnType("nvarchar(max)");
+                    b.HasKey("Id");
 
-                                    b.Property<int?>("InstitutionId")
-                                     .HasColumnType("int");
+                    b.ToTable("Tag");
+                });
```
Comment (niraymak) [inline Data/Migrations/ApplicationDbContextModelSnapshot.cs:359]: Not sure why this is here, can you check if this will cause any problems?

File: Repositories/ProjectCategoryRepository.cs [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+        public Task<ProjectCategory> GetProjectCategory(int projectId, int categoryId)
+        {
+            return DbSet.Where(p => p.Category.Id == categoryId && p.Project.Id == projectId)
+                        .FirstOrDefaultAsync();
+        }
+
+        /// <summary>
+        ///     Gets project category by given category id
+        /// </summary>
+        public Task<ProjectCategory> GetProjectCategory(int tagId)
```
Comment (niraymak) [inline Repositories/ProjectCategoryRepository.cs:69]: categoryId

File: Repositories/ProjectCategoryRepository.cs [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            return DbSet.Where(p => p.Category.Id == tagId)
```
Comment (niraymak) [inline Repositories/ProjectCategoryRepository.cs:71]: categoryId

File: Data/Migrations/ApplicationDbContextModelSnapshot.cs (hunk 17 above)
Comment (JVerbruggen) [reply Data/Migrations/ApplicationDbContextModelSnapshot.cs:359]: Dex seems to work fine after deletion of this file.. Should I push it? @niraymak

**Review by niraymak — CHANGES_REQUESTED** [review]
Thanks for the good work! 

Please look into my requests.

File: Data/Migrations/ApplicationDbContextModelSnapshot.cs (hunk 17 above)
Comment (JVerbruggen) [reply Data/Migrations/ApplicationDbContextModelSnapshot.cs:359]: https://www.learnentityframeworkcore.com/migrations/model-snapshot I Dont think we should though

**Comment by JVerbruggen** [comment]
Currently breaking postman:

```
  #  failure         detail                                                                             
                                                                                                        
 1.  AssertionError  Check if updated graduation date matches: 2021-03-12T09:20:22.310Z                 
     iteration: 3    expected '2021-03-12T09:20:22.31Z' to deeply equal '2021-03-12T09:20:22.310Z'      
                     at assertion:2 in test-script                                                      
                     inside "ACL / ConvertRegisteredToAlumni / ChangeExpectedGraduationDate-Registered" 
Error: Process completed with exit code 1.
```

File: Repositories/ProjectRepository.cs [hunk 20]
```diff
@@ -175,9 +179,24 @@ public override async Task<Project> FindAsync(int id)
...
+                project.User = RedactUser(project.User);
+                project.Likes = await GetDbSet<ProjectLike>()
+                                      .Where(p => p.LikedProject.Id == project.Id)
+                                      .ToListAsync();
+                project.Categories = await GetDbSet<ProjectCategory>()
+                                      .Include(p => p.Category)
+                                      .Where(p => p.Project.Id == project.Id)
+                                      .ToListAsync();
+            }
+            return await queryableProjects.ToListAsync();
```
Comment (BluRRayS) [inline Repositories/ProjectRepository.cs:199]: Not that it matters to much at this time but this could become really bad in the future performance wise then it might be better to create custom queries if the need arises

**Review by BluRRayS — APPROVED** [review]

File: Repositories/ProjectRepository.cs (hunk 20 above)
Comment (JVerbruggen) [reply Repositories/ProjectRepository.cs:199]: Can you make a suggestion on how you would improve this?

Comment (BluRRayS) [reply Repositories/ProjectRepository.cs:199]: I would make a function that would get a project with those specific relations that executes custom writen and optimized sql.  Since the documentation says the following:  "Raw SQL queries are also used if using a LINQ query is resulting in an inefficient SQL query." See https://docs.microsoft.com/en-us/ef/core/querying/raw-sql

Comment (JVerbruggen) [reply Repositories/ProjectRepository.cs:199]: Seems like a good idea. Maybe I'll pick it up myself in a follow-up issue, since it's a bit out of scope here. @rubenb994 what do you think? Should I make a new issue?

File: API/Controllers/ProjectController.cs [hunk 21]
```diff
@@ -792,6 +806,176 @@ public async Task<IActionResult> UnlikeProject(int projectId)
...
+        /// <param name="projectId"></param>
+        /// <param name="categoryId"></param>
+        /// <response code="200">Category was added to the project.</response>
+        /// <response code="401">User is not authorized to add the category to the project.</response>
+        /// <response code="404">Project or category was not found.</response>
+        /// <response code="409">Project is already categorized with the category.</response>
+        /// <returns></returns>
+        [HttpPost("category/{projectId}/{categoryId}")]
+        [Authorize]
+        [ProducesResponseType(StatusCodes.Status200OK)]
```
Comment (RubenFricke) [inline API/Controllers/ProjectController.cs:821]: Try to specify the return type if there is one. This improves the swagger quality.

```suggestion
[ProducesResponseType(typeof(ProjectCategoryResourceResult), StatusCodes.Status200OK)]
```

File: API/Controllers/ProjectController.cs [hunk 22]
```diff
@@ -792,6 +806,176 @@ public async Task<IActionResult> UnlikeProject(int projectId)
...
+        /// <param name="projectId"></param>
+        /// <param name="categoryId"></param>
+        /// <response code="200">Category was removed from the project.</response>
+        /// <response code="401">User is not authorized to remove the category from the project.</response>
+        /// <response code="404">Project or category was not found.</response>
+        /// <response code="409">Project is not categorized with the category.</response>
+        /// <returns></returns>
+        [HttpDelete("category/{projectId}/{categoryId}")]
+        [Authorize]
+        [ProducesResponseType(StatusCodes.Status200OK)]
```
Comment (RubenFricke) [inline API/Controllers/ProjectController.cs:907]: Same here, specify return type

File: CHANGELOG.md [hunk 23]
```diff
@@ -15,6 +15,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Graduation system for users whose FHICT account becomes unavailable. [#287](https://github.com/DigitalExcellence/dex-backend/issues/287)
 - Added wizard to automatically import project with a public and an oauth flow - [326](https://github.com/DigitalExcellence/dex-backend/issues/326)
 
+- Added tags - [362](https://github.com/DigitalExcellence/dex-backend/issues/362)
```
Comment (RubenFricke) [inline CHANGELOG.md:18]:
```suggestion
- Added tags - [#362](https://github.com/DigitalExcellence/dex-backend/issues/362)
```

**Review by RubenFricke — CHANGES_REQUESTED** [review]
A few small notes I found (nothing too fancy). Furthermore, I had one small question. Wouldn't it be better if you could add categories when you create the project? Now you have to create the project and update the project to add categories. I think it would make more sense if you call add categories when you create the projects (if you disagree, let me know :wink: ).

The code looks really good again, I don't really have any changes.

**Comment by JVerbruggen** [comment]
> A few small notes I found (nothing too fancy). Furthermore, I had one small question. Wouldn't it be better if you could add categories when you create the project? Now you have to create the project and update the project to add categories. I think it would make more sense if you call add categories when you create the projects (if you disagree, let me know 😉 ).
> 
> The code looks really good again, I don't really have any changes.

Should've thought about that. I think I should add this to the project controller (create & update project) so that you can provide categories with it immediately. I also think that we should keep the current functionality, should you want to add categories to a project without editing the project.

File: Repositories/ProjectRepository.cs (hunk 20 above)
Comment (RubenFricke) [reply Repositories/ProjectRepository.cs:199]: @JVerbruggen For me, you could either resolve it in this issue because this method is used in this issue scope as well, or make a new issue if you would like to tackle all methods in the repository layer to make sure the performance is ok. (I assume you meant to tag me by the way :smile: ?)

File: Models/Project.cs [hunk 24]
```diff
@@ -75,6 +75,9 @@ public Project()
 
         public bool InstitutePrivate { get; set; }
 
+        public List<ProjectCategory> Categories { get; set; } = new List<ProjectCategory>();
```
Comment (TimSnoek123) [inline Models/Project.cs:78]: This new should be deleted I think

File: Repositories/ProjectCategoryRepository.cs [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+        /// </summary>
+        /// <param name="dbContext"></param>
+        public ProjectCategoryRepository(DbContext dbContext) :
+            base(dbContext) { }
+
+        /// <summary>
+        ///     Gets project category by given project and category id
+        /// </summary>
+        public Task<ProjectCategory> GetProjectCategory(int projectId, int categoryId)
+        {
```
Comment (TimSnoek123) [inline Repositories/ProjectCategoryRepository.cs:61]: You can put the lambda in the firstOrDefault here instead of the where  I think

File: Repositories/ProjectCategoryRepository.cs [hunk 26]
```diff
@@ -0,0 +1,75 @@
...
+            return DbSet.Where(p => p.Category.Id == categoryId && p.Project.Id == projectId)
+                        .FirstOrDefaultAsync();
+        }
+
+        /// <summary>
+        ///     Gets project category by given category id
+        /// </summary>
+        public Task<ProjectCategory> GetProjectCategory(int categoryId)
```
Comment (TimSnoek123) [inline Repositories/ProjectCategoryRepository.cs:69]: Same here

File: Services/Services/ProjectCategoryService.cs [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+        {
+            return Repository.GetProjectCategory(categoryId);
+        }
+
+        /// <summary>
+        ///     Update project categories by given project and categories
+        /// </summary>
+        public async Task UpdateProjectCategories(Project project, IEnumerable<Category> categories)
+        {
+            IEnumerable<ProjectCategory> allProjectCategories = await Repository.GetAll();
```
Comment (TimSnoek123) [inline Services/Services/ProjectCategoryService.cs:95]: I think its better to use a get by projectId here instead of loading all ProjectCategories in memory

Comment (JVerbruggen) [reply Services/Services/ProjectCategoryService.cs:95]: Nuances.. Maybe for a next issue?

File: Repositories/ProjectCategoryRepository.cs (hunk 26 above)
Comment (JVerbruggen) [reply Repositories/ProjectCategoryRepository.cs:69]: Nuances.. Maybe for a next issue?

File: Repositories/ProjectCategoryRepository.cs (hunk 25 above)
Comment (JVerbruggen) [reply Repositories/ProjectCategoryRepository.cs:61]: Nuances.. Maybe for a next issue?

File: Models/Project.cs (hunk 24 above)
Comment (JVerbruggen) [reply Models/Project.cs:78]: Why?

**Review by TimSnoek123 — APPROVED** [review]

File: API/Controllers/ProjectController.cs [hunk 28]
```diff
@@ -1058,6 +1104,176 @@ public async Task<IActionResult> UnlinkInstitutionFromProjectAsync(int projectId
...
+            }
+
+            User user = await HttpContext.GetContextUser(userService)
+                                         .ConfigureAwait(false);
+            bool isAllowed = await authorizationHelper.UserIsAllowed(user,
+                                                                     nameof(Defaults.Scopes.AdminProjectWrite),
+                                                                     nameof(Defaults.Scopes.InstitutionProjectWrite),
+                                                                     project.UserId);
+
+            if(project.UserId != user.Id && !isAllowed)
```
Comment (niraymak) [inline API/Controllers/ProjectController.cs:1246]: Unauthorized should return if one or both conditions are false, right?

```suggestion
            if(!(project.UserId == user.Id || isAllowed))
```

File: Models/Project.cs (hunk 24 above)
Comment (niraymak) [reply Models/Project.cs:78]: Why not? I don't see it at one of the other properties?

File: Data/Migrations/ApplicationDbContextModelSnapshot.cs [hunk 29]
```diff
@@ -365,6 +365,21 @@ protected override void BuildModel(ModelBuilder modelBuilder)
...
 
+            modelBuilder.Entity("Models.Category", b =>
+                {
+                    b.Property<int>("Id")
+                        .ValueGeneratedOnAdd()
+                        .HasColumnType("int")
+                        .HasAnnotation("SqlServer:ValueGenerationStrategy", SqlServerValueGenerationStrategy.IdentityColumn);
+
+                    b.Property<string>("Name")
+                        .HasColumnType("nvarchar(max)");
```
Comment (niraymak) [inline Data/Migrations/ApplicationDbContextModelSnapshot.cs:376]: Maybe we shouldn't make this column nvarchar(max) but nvarchar(100)

File: API/Controllers/ProjectController.cs [hunk 30]
```diff
@@ -1058,6 +1104,176 @@ public async Task<IActionResult> UnlinkInstitutionFromProjectAsync(int projectId
...
```
Comment (niraymak) [inline API/Controllers/ProjectController.cs:1160]: Unauthorized should return if one or both conditions are false, right?

```suggestion
            if(!(project.UserId == user.Id || isAllowed))
```

**Review by niraymak — CHANGES_REQUESTED** [review]
Good work! 👍 

I tried to update the categories of an existing project but something went wrong mapping the resource to an actual project. Maybe just to be sure you can also add a check for this in the integration tests.

Also found an issue for the authorization check.

Maybe since you have to make some changes you can also fix Tim's feedback.

File: API/Controllers/ProjectController.cs (hunk 28 above)
Comment (JVerbruggen) [reply API/Controllers/ProjectController.cs:1246]: I think the correction you gave is very confusing. I read it as the following:

**If project userid is not user id AND not allowed, then fail.**
Yours, I would read like this:
**if not project user id is user id OR is allowed**
This makes it very hard to read to me.

In essence, it does the same. Look:

Mine:
userId = userId | isAllowed | result
---|---|---
false | false | true
false | true | false
true | false | false
true | true | false

Yours:
userId = userId | isAllowed | result
---|---|---
true | true | false
true | false | false
false | true | false
false | false | true

It does the same, i just think yours is harder to read.

Comment (niraymak) [reply API/Controllers/ProjectController.cs:1246]: Correct me if I'm wrong but I think in your case both conditions should meet before it would enter the statement.
userId is not equal **AND** isAllowed = false. 

I think if only one of them meets, it should enter the statement.

@Maxvanhattum maybe second opinion?

**Review by niraymak — CHANGES_REQUESTED** [review]
Results for request for projectId 59:
image

Tried to update categories:
image

Project categories not updated succesfully.
image

**Review by niraymak — APPROVED** [review]
Thanks for your implementation :) Good work! 💯 

Only had some doubts if it should be possible updating the categories using the UpdateProject endpoint. Spoke with @RubenFricke about it and he told me it shouldn't be a big issue for the frontend.

**Review by niraymak — APPROVED** [review]

**Review by niraymak — APPROVED** [review]

**Review by RubenFricke — APPROVED** [review]
Everything works flawless. Thanks for the pull request!

**Review by TimSnoek123 — APPROVED** [review]
Looks good 👍
